File: planning-tools/src/main/kotlin/com/pilothelper/model/Weather.kt
package com.pilothelper.model

import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.PrimitiveKind
import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder

/**
 * Data received from AviationWeatherGov API: https://aviationweather.gov/data/api/#/Data/dataMetars
 */
@Serializable
data class Weather(
    val metar_id: Int,
    val icaoId: String,
    val receiptTime: String,
    val obsTime: Int,
    val reportTime: String,
    val temp: Int,
    val dewp: Int,
    val wdir: Int,
    val wspd: Int,
    val wgst: Int?,
    @Serializable(with = FloatOrStringAsStringSerializer::class)
    val visib: String,
    val altim: Int,
    val slp: String?,
    val qcField: Int,
    val wxString: String?,
    val presTend: String?,
    val maxT: String?,
    val minT: String?,
    val maxT24: String?,
    val minT24: String?,
    val precip: String?,
    val pcp3hr: String?,
    val pcp6hr: String?,
    val pcp24hr: String?,
    val snow: String?,
    val vertVis: String?,
    val metarType: String,
    val rawOb: String,
    val mostRecent: Int,
    val lat: Float,
    val lon: Float,
    val elev: Int,
    val prior: Int,
    val name: String,
    val clouds: List<Cloud>,
    // ...
) {
    @Serializable
    data class Cloud(
        val cover: String,
        val base: Int?
    )
}


object FloatOrStringAsStringSerializer : KSerializer<String> {
    override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor("FloatOrString", PrimitiveKind.STRING)

    override fun serialize(encoder: Encoder, value: String) {
        encoder.encodeString(value)
    }

    override fun deserialize(decoder: Decoder): String {
        return try {
            decoder.decodeString()
        } catch (e: Exception) {
            decoder.decodeFloat().toString()
        }
    }
}

File: frontend/src/shared/models/Weather.ts
export class Cloud {
  constructor(
    public cover: string,
    public base: number | null
  ) {
  }
}

export class Weather {
  constructor(
    public metar_id: number,
    public icaoId: string,
    public receiptTime: string,
    public obsTime: number,
    public reportTime: string,
    public temp: number,
    public dewp: number,
    public wdir: number,
    public wspd: number,
    public wgst: string | null,
    public visib: string,
    public altim: number,
    public slp: string | null,
    public qcField: number,
    public wxString: string | null,
    public presTend: string | null,
    public maxT: string | null,
    public minT: string | null,
    public maxT24: string | null,
    public minT24: string | null,
    public precip: string | null,
    public pcp3hr: string | null,
    public pcp6hr: string | null,
    public pcp24hr: string | null,
    public snow: string | null,
    public vertVis: string | null,
    public metarType: string,
    public rawOb: string,
    public mostRecent: number,
    public lat: number,
    public lon: number,
    public elev: number,
    public prior: number,
    public name: string,
    public clouds: Cloud[]
  ) {
  }
}
